File: kqsp_cli.js
// kqsp_cli.js - Node.js CLI for Kazan's Quick Share Protocol

// Import necessary modules
const {
    Peer
} = require('peerjs'); // PeerJS library for WebRTC
const wrtc = require('wrtc'); // Node.js WebRTC implementation
const readline = require('readline'); // For reading user input from console
const crypto = require('crypto'); // For SHA256 hashing (group key)
const fs = require('fs'); // For potential file operations
const os = require('os'); // For potential OS-specific features
const process = require('process'); // To access command line arguments

// --- Configuration & Global State ---
let MY_PEER_ID = null;
let MY_DISPLAY_ADDR = null;
let peer = null; // PeerJS instance
let group_key = null;
const connections = new Map(); // Map peerId -> DataConnection

// --- K(addr) Generation --- (Simplified version)
function generateKAddr() {
    const parts = Array.from({
        length: 4
    }, () => Math.floor(Math.random() * 256));
    const display = parts.join('.');
    // PeerJS IDs need to be strings, often alphanumeric. Let's use a simpler format for now.
    // A robust version might use UUIDs or ensure uniqueness via the signaling server.
    const peerId = `kqsp-cli-${parts.join('-')}-${Date.now().toString(36).slice(-4)}`;
    return {
        display: `K(${display})`,
        peerId
    };
}

// --- Group Key & Encryption --- (Same logic as web/python)
function updateGroupKey() {
    const peerIds = sorted([MY_PEER_ID, ...connections.keys()]);
    const hasher = crypto.createHash('sha256');
    hasher.update(JSON.stringify(peerIds));
    group_key = hasher.digest();
    // console.log(`[Debug] Group key updated based on: ${peerIds.join(', ')}`);
}

function xorCrypt(dataBuffer, keyBuffer) {
    if (!keyBuffer) {
        return dataBuffer; // No encryption if key is not set
    }
    const keyLen = keyBuffer.length;
    const result = Buffer.alloc(dataBuffer.length);
    for (let i = 0; i < dataBuffer.length; i++) {
        result[i] = dataBuffer[i] ^ keyBuffer[i % keyLen];
    }
    return result;
}

// --- PeerJS Event Handlers ---
function initializePeer(peerId, targetPeerId = null) { // <-- Add targetPeerId argument
    // IMPORTANT: Configure this to match your web version's signaling server!
    const peerJsOptions = {
        host: '0.peerjs.com', // Default public PeerJS server
        port: 443,
        path: '/',
        secure: true,
        // key: 'peerjs', // Default key for public server
        debug: 2, // 0: Errors, 1: Warnings, 2: Info, 3: Verbose
        wrtc: wrtc // Provide the wrtc implementation for Node.js
    };

    peer = new Peer(peerId, peerJsOptions);

    peer.on('open', (id) => {
        MY_PEER_ID = id;
        console.log(`[System] PeerJS connection open. Your Peer ID: ${id}`);
        console.log(`          Your K(addr): ${MY_DISPLAY_ADDR}`);
        updateGroupKey(); // Initial key

        // Auto-connect if targetPeerId was provided via command line
        if (targetPeerId) {
            console.log(`[System] Auto-connecting to ${targetPeerId} from command line...`);
            const conn = peer.connect(targetPeerId, {
                reliable: true,
            });
            setupConnectionHandlers(conn);
        }

        promptUser(); // Start accepting user input
    });

    peer.on('connection', (conn) => {
        console.log(`[System] Incoming connection from ${conn.peer}`);
        setupConnectionHandlers(conn);
    });

    peer.on('disconnected', () => {
        console.log('[System] Disconnected from signaling server. Attempting to reconnect...');
        // PeerJS attempts reconnection automatically
        // Consider adding manual reconnection logic if needed
    });

    peer.on('close', () => {
        console.log('[System] Peer connection closed permanently.');
        // Handle cleanup if necessary
    });

    peer.on('error', (err) => {
        console.error(`[System] PeerJS Error: ${err.type} - ${err.message}`);
        // Handle specific errors (e.g., 'network', 'server-error', 'unavailable-id')
        if (err.type === 'unavailable-id') {
            console.error(`[System] Peer ID ${MY_PEER_ID} is already taken. Please restart.`);
            process.exit(1);
        }
        // Consider attempting to reconnect or exit based on error type
    });
}

function setupConnectionHandlers(conn) {
    conn.on('open', () => {
        console.log(`[System] Data connection established with ${conn.peer}`);
        connections.set(conn.peer, conn);
        updateGroupKey();
        // Optionally send a welcome message or request info
    });

    conn.on('data', (data) => {
        try {
            // PeerJS handles JSON serialization/deserialization if specified, but let's be explicit
            const payload = typeof data === 'string' ? JSON.parse(data) : data;

            if (payload.type === 'text') {
                // Data might be ArrayBuffer, Blob, string, etc. Need to handle Buffer for XOR
                // Assuming text is sent latin1 encoded buffer string from other clients
                const encryptedBuffer = Buffer.from(payload.text, 'latin1');
                const decryptedBuffer = xorCrypt(encryptedBuffer, group_key);
                console.log(`${payload.from}: ${decryptedBuffer.toString('utf-8')}`);
            } else {
                console.log(`[System] Received unhandled message type '${payload.type}' from ${conn.peer}`);
            }
        } catch (e) {
            console.error(`[System] Error processing data from ${conn.peer}:`, e);
            console.error(`[System] Raw data received:`, data);
        }
        promptUser(); // Re-display prompt after message
    });

    conn.on('close', () => {
        console.log(`[System] Connection closed with ${conn.peer}`);
        connections.delete(conn.peer);
        updateGroupKey();
    });

    conn.on('error', (err) => {
        console.error(`[System] Connection error with ${conn.peer}: ${err.message}`);
        connections.delete(conn.peer);
        updateGroupKey();
    });
}

// --- Message Sending ---
function sendTextMessage(text) {
    if (connections.size === 0) {
        console.log('[System] No active connections to send message.');
        return;
    }
    if (!group_key) {
        console.log('[System] Group key not yet established. Cannot send message.');
        return;
    }

    const textBuffer = Buffer.from(text, 'utf-8');
    const encryptedBuffer = xorCrypt(textBuffer, group_key);

    const payload = {
        type: 'text',
        from: MY_DISPLAY_ADDR,
        text: encryptedBuffer.toString('latin1') // Encode buffer as latin1 string for transport
    };

    const messageStr = JSON.stringify(payload);
    let sentToAny = false;

    connections.forEach((conn, peerId) => {
        if (conn.open) {
            try {
                conn.send(messageStr);
                sentToAny = true;
            } catch (e) {
                console.error(`[System] Failed to send to ${peerId}: ${e.message}`);
            }
        } else {
            console.log(`[System] Connection to ${peerId} is not open.`);
        }
    });

    if (sentToAny) {
        console.log(`You: ${text}`); // Show own message
    }
}

// --- Command Line Interface (CLI) --- //
const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout,
    prompt: '> '
});

function promptUser() {
    rl.prompt();
}

rl.on('line', (line) => {
    const input = line.trim();
    if (input.startsWith('/connect ')) {
        const targetPeerId = input.substring('/connect '.length).trim();
        if (targetPeerId && peer) {
            console.log(`[System] Attempting to connect to ${targetPeerId}...`);
            const conn = peer.connect(targetPeerId, {
                reliable: true, // Use reliable SCTP
                // serialization: 'json' // Let PeerJS handle JSON
            });
            setupConnectionHandlers(conn);
        } else {
            console.log('[System] Invalid /connect command. Usage: /connect <target_peer_id>');
        }
    } else if (input === '/peers') {
        if (connections.size > 0) {
            console.log('[System] Connected peers:');
            connections.forEach((conn, peerId) => {
                console.log(` - ${peerId} (${conn.open ? 'open' : 'pending/closed'})`);
            });
        } else {
            console.log('[System] No peers connected.');
        }
    } else if (input === '/myid') {
        console.log(`[System] Your Peer ID: ${MY_PEER_ID}`);
        console.log(`          Your K(addr): ${MY_DISPLAY_ADDR}`);
    } else if (input === '/quit') {
        console.log('[System] Exiting...');
        rl.close();
        if (peer) {
            peer.destroy(); // Close connections and disconnect from signaling server
        }
        process.exit(0);
    } else if (input.startsWith('/')) {
        console.log(`[System] Unknown command: ${input.split(' ')[0]}`);
        console.log('[System] Available commands: /connect <peer_id>, /peers, /myid, /quit');
    } else if (input) {
        sendTextMessage(input);
    }
    promptUser(); // Always re-prompt unless quitting
});

rl.on('close', () => {
    console.log(' [System] Readline closed.Exiting.');
    if (peer) {
        peer.destroy();
    }
    process.exit(0);
});

// --- Initialization ---
function main(targetPeerId = null) { // <-- Add targetPeerId argument
    const {
        display,
        peerId
    } = generateKAddr();
    MY_DISPLAY_ADDR = display;
    // Use generated ID or allow override via args later?
    initializePeer(peerId, targetPeerId); // <-- Pass targetPeerId
}

// Helper for sorting keys
function sorted(arr) {
    return [...arr].sort();
}

// --- Argument Parsing & Start --- //
function parseArgs() {
    const args = process.argv.slice(2); // Skip 'node' and script path
    let targetPeerId = null;
    const connectIndex = args.indexOf('--connect');

    if (connectIndex !== -1 && args.length > connectIndex + 1) {
        targetPeerId = args[connectIndex + 1];
        // Remove K() or K prefix if present, as PeerJS IDs don't use it
        if (targetPeerId.startsWith('K(') && targetPeerId.endsWith(')')) {
            targetPeerId = targetPeerId.substring(2, targetPeerId.length - 1);
        } else if (targetPeerId.startsWith('K')) {
            targetPeerId = targetPeerId.substring(1);
        }
        // Basic validation: PeerJS IDs usually don't look like IP addresses
        // This is a weak check, PeerJS IDs can be varied.
        // A better approach might be needed if K(addr) format needs direct mapping.
        // For now, we assume the user provides the actual PeerJS ID.
        console.log(`[System] Target Peer ID from args: ${targetPeerId}`);
    }
    return targetPeerId;
}

const targetPeerIdFromArgs = parseArgs();
main(targetPeerIdFromArgs); // <-- Pass parsed target ID to main

// Remove the duplicate call below
// main();

File: kqsp_cli.py
import asyncio
import json
import random
import hashlib
import sys
import ssl
import logging
import os
from aiortc import RTCIceCandidate, RTCPeerConnection, RTCSessionDescription, MediaStreamTrack
from aiortc.contrib.signaling import BYE, add_signaling_arguments, create_signaling

# --- K(addr) Generation & Global State ---
def generate_k_addr():
    """Generates a unique K(addr) ID and display string."""
    parts = [random.randint(0, 255) for _ in range(4)]
    display = '.'.join(map(str, parts))
    peer_id = '-'.join(map(str, parts))
    return display, peer_id

MY_DISPLAY_ADDR, MY_PEER_ID = generate_k_addr()
ROOT = os.path.dirname(__file__)
cert_file = os.path.join(ROOT, "cert.pem") # Placeholder for potential cert
key_file = os.path.join(ROOT, "key.pem")   # Placeholder for potential key

pcs = set() # Set of active RTCPeerConnection objects
group_key = None
message_queue = asyncio.Queue() # Async queue for received messages
stop_event = asyncio.Event()

# --- Group Key & Encryption (Same as before) ---
def update_group_key(connected_peer_ids):
    """Updates the group key based on connected peers."""
    global group_key
    peer_ids = sorted([MY_PEER_ID] + list(connected_peer_ids))
    hasher = hashlib.sha256()
    hasher.update(json.dumps(peer_ids).encode('utf-8'))
    group_key = hasher.digest()
    # print(f"[Debug] Group key updated based on: {peer_ids}")

def xor_crypt(data_bytes, key_bytes):
    """Encrypts/decrypts data using XOR with the key."""
    if not key_bytes:
        return data_bytes # No encryption if key is not set
    key_len = len(key_bytes)
    return bytes([b ^ key_bytes[i % key_len] for i, b in enumerate(data_bytes)])

# --- WebRTC Data Channel Handling ---
async def handle_data_channel(channel, peer_id):
    """Handles messages received on a data channel."""
    await message_queue.put(f"[System] Data channel '{channel.label}' created with {peer_id}")
    update_group_key({pc.remoteDescription.sdp.split('o=')[1].split(' ')[0] for pc in pcs if pc.remoteDescription}) # Update key on new channel

    @channel.on("message")
    async def on_message(message):
        try:
            # Assume message is JSON payload as string
            payload = json.loads(message)
            if payload.get('type') == 'text':
                encrypted_text_bytes = payload['text'].encode('latin-1')
                decrypted_bytes = xor_crypt(encrypted_text_bytes, group_key)
                await message_queue.put(f"{payload['from']}: {decrypted_bytes.decode('utf-8')}")
            # Add handlers for 'file', 'audio' later
            else:
                await message_queue.put(f"[System] Received unknown message type from {peer_id}: {payload.get('type')}")
        except (json.JSONDecodeError, UnicodeDecodeError, KeyError, AttributeError) as e:
            await message_queue.put(f"[System] Error processing message from {peer_id} on channel {channel.label}: {e}")
            # Log raw message for debugging if not JSON
            if isinstance(message, str):
                 await message_queue.put(f"[System] Raw data: {message[:100]}...")
            else:
                 await message_queue.put(f"[System] Received non-string data: {type(message)}")

    @channel.on("close")
    async def on_close():
        await message_queue.put(f"[System] Data channel '{channel.label}' closed with {peer_id}")
        # Find the peer connection associated with this channel to remove if needed
        # This part is tricky as channel doesn't directly link back to pc easily
        # We might need to manage connections differently

# --- Main Application Logic ---
async def run(pc, signaling, role):
    """Main coroutine for handling a single peer connection."""
    connected_peer_ids = set()

    def get_peer_id(pc_obj):
        if pc_obj and pc_obj.remoteDescription:
            # Attempt to extract peer ID from SDP (this is fragile)
            try:
                return pc_obj.remoteDescription.sdp.split('o=')[1].split(' ')[0]
            except IndexError:
                return "unknown-peer"
        return "unknown-peer"

    async def send_ping():
        # Keepalive or other periodic tasks if needed
        pass

    @pc.on("datachannel")
    async def on_datachannel(channel):
        peer_id = get_peer_id(pc)
        asyncio.create_task(handle_data_channel(channel, peer_id))

    @pc.on("connectionstatechange")
    async def on_connectionstatechange():
        peer_id = get_peer_id(pc)
        await message_queue.put(f"[System] Connection state with {peer_id} is {pc.connectionState}")
        if pc.connectionState == "failed" or pc.connectionState == "closed" or pc.connectionState == "disconnected":
            await signaling.send(BYE)
            if pc in pcs:
                pcs.discard(pc)
            connected_peer_ids.discard(peer_id)
            update_group_key(connected_peer_ids)
            await pc.close()
        elif pc.connectionState == "connected":
             connected_peer_ids.add(peer_id)
             update_group_key(connected_peer_ids)

    @pc.on("track")
    def on_track(track):
        # KQSP doesn't use media tracks currently, but handle gracefully
        @track.on("ended")
        async def on_ended():
             await message_queue.put(f"[System] Track {track.kind} ended")

    # Connect signaling
    await signaling.connect()

    if role == "offer":
        # Create data channel
        channel = pc.createDataChannel("kqsp-chat")
        peer_id = "offering-peer" # Placeholder until connected
        asyncio.create_task(handle_data_channel(channel, peer_id))

        # Send offer
        await pc.setLocalDescription(await pc.createOffer())
        await signaling.send(pc.localDescription)
        await message_queue.put("[System] Sent offer...")

    # Consume signaling
    while True:
        try:
            obj = await signaling.receive()

            if isinstance(obj, RTCSessionDescription):
                await pc.setRemoteDescription(obj)
                peer_id = get_peer_id(pc)
                await message_queue.put(f"[System] Received remote description from {peer_id}")

                if obj.type == "offer":
                    # Create data channel if answering
                    # Note: on_datachannel handles channel creation initiated by remote peer
                    # Send answer
                    await pc.setLocalDescription(await pc.createAnswer())
                    await signaling.send(pc.localDescription)
                    await message_queue.put("[System] Sent answer...")
            elif isinstance(obj, RTCIceCandidate):
                await pc.addIceCandidate(obj)
                # await message_queue.put("[System] Added ICE candidate") # Too verbose
            elif obj is BYE:
                await message_queue.put("[System] Received BYE, closing connection")
                break
            else:
                await message_queue.put(f"[System] Received unknown signaling object: {type(obj)}")
        except asyncio.CancelledError:
            break
        except Exception as e:
            await message_queue.put(f"[System] Error during signaling: {e}")
            break

    await message_queue.put("[System] Signaling loop ended.")

# --- User Input and Message Sending ---
async def consume_user_input():
    """Handles user input from stdin."""
    loop = asyncio.get_event_loop()
    while not stop_event.is_set():
        try:
            # Use run_in_executor for blocking input()
            user_input = await loop.run_in_executor(
                None, lambda: input("Enter message or command (/quit): ")
            )
            if user_input == '/quit':
                await message_queue.put("[System] Quitting...")
                stop_event.set()
                break
            elif user_input.startswith('/'):
                await message_queue.put(f"[System] Unknown command: {user_input.split(' ')[0]}")
            elif user_input:
                await send_cli_message(user_input)
        except (EOFError, KeyboardInterrupt):
            await message_queue.put("\n[System] Input interrupted. Quitting...")
            stop_event.set()
            break
        except Exception as e:
            await message_queue.put(f"\n[System] Error reading input: {e}")
            await asyncio.sleep(1)

async def send_cli_message(text):
    """Encrypts and sends a text message to all connected peers via data channels."""
    if not pcs:
        await message_queue.put("[System] No active connections to send message.")
        return
    if not group_key:
        await message_queue.put("[System] Group key not yet established. Cannot send message.")
        return

    encrypted_bytes = xor_crypt(text.encode('utf-8'), group_key)
    payload = {
        'type': 'text',
        'from': f"K({MY_DISPLAY_ADDR})",
        'text': encrypted_bytes.decode('latin-1') # Use latin-1 to preserve byte values
    }
    message_str = json.dumps(payload)

    sent_to_any = False
    failed_pcs = set()
    for pc in pcs:
        # Find the 'kqsp-chat' data channel (or similar)
        channel = next((c for c in pc.sctp.dataChannels if c.label == 'kqsp-chat' and c.readyState == 'open'), None)
        if channel:
            try:
                channel.send(message_str)
                sent_to_any = True
            except Exception as e:
                peer_id = get_peer_id(pc)
                await message_queue.put(f"[System] Failed to send to {peer_id}: {e}")
                failed_pcs.add(pc)
        else:
            peer_id = get_peer_id(pc)
            await message_queue.put(f"[System] No open 'kqsp-chat' channel found for {peer_id}")
            # Consider closing this pc if channel is missing/closed?
            # failed_pcs.add(pc)

    if sent_to_any:
         await message_queue.put(f"You: {text}") # Show own message
    else:
         await message_queue.put("[System] Message could not be sent to any peer.")

    # Clean up failed connections
    # for pc in failed_pcs:
    #     if pc in pcs:
    #         await message_queue.put(f"[System] Closing failed connection {get_peer_id(pc)}")
    #         pcs.discard(pc)
    #         await pc.close()

# --- Message Printing --- #
async def print_messages():
    """Prints messages from the async queue."""
    while not stop_event.is_set():
        try:
            msg = await asyncio.wait_for(message_queue.get(), timeout=0.1)
            # Simple print for now, could clear line like before if needed
            print(msg)
            message_queue.task_done()
        except asyncio.TimeoutError:
            continue
        except Exception as e:
            print(f"\n[System] Error printing message: {e}")

# --- Main Execution --- #
async def main(args):
    print("--- Kazan's Quick Share Protocol (CLI - WebRTC) ---")
    print(f"Your PeerID: {MY_PEER_ID}")
    print(f"Signaling Server: {args.signaling_host}:{args.signaling_port}")
    print("Connecting to signaling server...")
    print("--------------------------------------------------")

    # Create signaling instance
    # Note: This basic example uses aiortc's built-in TCP signaling.
    # For web compatibility, a WebSocket signaling server (like PeerJS server or custom) is needed.
    signaling = create_signaling(args)

    # Create peer connection
    pc = RTCPeerConnection()
    pcs.add(pc)

    # Start background tasks
    print_task = asyncio.create_task(print_messages())
    input_task = asyncio.create_task(consume_user_input())

    # Run main connection logic
    try:
        await run(pc=pc, signaling=signaling, role=args.role)
    except Exception as e:
        print(f"[System] Main execution error: {e}")
    finally:
        print("[System] Cleaning up...")
        stop_event.set()

        # Close signaling and connections
        await signaling.close()
        coros = [pc.close() for pc in pcs]
        await asyncio.gather(*coros)
        pcs.clear()

        # Cancel background tasks
        input_task.cancel()
        print_task.cancel()
        try:
            await input_task
        except asyncio.CancelledError:
            pass
        try:
            await print_task
        except asyncio.CancelledError:
            pass

    print("[System] Exited.")

if __name__ == "__main__":
    # Setup logging
    logging.basicConfig(level=logging.INFO)
    logging.getLogger("aiortc").setLevel(logging.WARNING)
    logging.getLogger("aioice").setLevel(logging.WARNING)

    # Add signaling arguments (host, port, role)
    parser = add_signaling_arguments()
    # Add any KQSP specific arguments here if needed
    args = parser.parse_args()

    # SSL context (optional, depends on signaling server)
    # ssl_context = ssl.SSLContext(ssl.PROTOCOL_TLS_CLIENT)
    # ssl_context.load_verify_locations(cert_file)

    try:
        asyncio.run(main(args))
    except KeyboardInterrupt:
        pass